uniffi_bindgen/src/bindings/kotlin/templates/ObjectTemplate.kt

```diff
@@ -267,20 +283,52 @@ open class {{ impl_class_name }}: Disposable, AutoCloseable, {{ interface_name }
 {% include "CallbackInterfaceImpl.kt" %}
 {%- endif %}
 
+{%- let trait_impl=format!("uniffiCallbackInterface{}", name) %}
+
+{%- if obj.has_callback_interface() %}
 /**
  * @suppress
  */
-public object {{ ffi_converter_name }}: FfiConverter<{{ type_name }}, Pointer> {
-    {%- if obj.has_callback_interface() %}
+public object {{ ffi_converter_name }}: FfiConverterRustBuffer<{{ type_name }}> {
     internal val handleMap = UniffiHandleMap<{{ type_name }}>()
-    {%- endif %}
+    internal val langIndex: Int = {{ trait_impl }}.register(UniffiLib.INSTANCE)
+
+    override fun read(buf: ByteBuffer): {{ type_name }} {
+        // The Rust code always writes pointers as 8 bytes, and will
+        // fail to compile if they don't fit.
+        val handle = buf.getLong()
+        val langIndex = buf.getInt()
+        if (handleMap.has(handle)) {
+            return handleMap.get(handle)
+        } else {
+            return {{ impl_class_name }}(Pointer(handle))
+        }
+    }
+
+    override fun allocationSize(value: {{ type_name }}) = 12UL
 
+    fun lowerPointer(pointer: Pointer): RustBuffer.ByValue =
+        RustBuffer.write(12UL) { bbuf ->
+            bbuf.putLong(Pointer.nativeValue(pointer))
+            bbuf.putInt(langIndex)
```

Comment:
This is how we have to handle it with the current handle system.  I'm hoping to implement something like #1823, which would allow us to pack both the pointer and the `langIndex` into a 64-bit handle.

This means creating some sort of tagged pointer.  The simplest version of that would be using the lower bits for the langIndex.  We know that `Arc` stores a word ref count, so on a 32-bit system it's aligned to at least `4` bytes.  That gives us 2 bits to use, which could support 3 foreign languages plus Rust, do you think that would be enough?

Reply:
3 languages at the same time sounds like it should be enough.

Just for the sake of not boxing ourselves into a corner: I should imagine if we want more than 3 then we'd need to move to 64bits for everything, which also sounds do-able, if absolutely necessary.